src/PowerShellEditorServices.Protocol/Server/LanguageServer.cs

```diff
@@ -1234,48 +1236,71 @@ public static string GetDecoratedSymbolName(SymbolReference symbolReference)
             await requestContext.SendResult(result);
         }
 
+        // Since the NamedPipeConnectionInfo type is only available in 5.1+
+        // we have to use reflection to support older version of PS.
+        // This code only lives in the v1.X of the extension.
+        // The 2.x version of the code can be found here:
+        // https://github.com/PowerShell/PowerShellEditorServices/pull/881
+        private static Type _namedPipeConnectionInfoType = Type.GetType("System.Management.Automation.Runspaces.NamedPipeConnectionInfo, System.Management.Automation");
+        private static MethodInfo _runspaceFactoryCreateRunspaceMethod = typeof(RunspaceFactory)
+                .GetMethod("CreateRunspace", new Type[] { _namedPipeConnectionInfoType });
+        private Runspace GetRemoteRunspace(int pid)
+        {
+            var namedPipeConnectionInfoInstance = Activator.CreateInstance(
+                _namedPipeConnectionInfoType,
+                pid);
+            return _runspaceFactoryCreateRunspaceMethod.Invoke(null, new [] { namedPipeConnectionInfoInstance }) as Runspace;
```

Comment:
I'd go with something like this:

```csharp
class LanguageServer
{
    private static readonly Func<int, RunspaceConnectionInfo> s_createNamedPipeRunspace;

    static LanguageServer()
    {
        s_createNamedPipeRunspace = CreateNamedPipeCtorDelegate();

        if (s_createNamedPipeRunspace == null)
        {
            s_createNamedPipeRunspace = new Func<int, RunspaceConnectionInfo>(DefaultNamedPipeCtorDelegate);
        }
    }

    private static Runspace GetRemoteRunspace(int pid)
        => RunspaceFactory.CreateRunspace(s_createNamedPipeRunspace(pid));

    private static RunspaceConnectionInfo DefaultNamedPipeCtorDelegate(int pid)
        => throw new NotSupportedException();

    private static Func<int, RunspaceConnectionInfo> CreateNamedPipeCtorDelegate()
    {
        Type namedPipeConnectionInfoType = typeof(PSObject).Assembly
            .GetType("System.Management.Automation.Runspaces.NamedPipeConnectionInfo");

        if (namedPipeConnectionInfoType == null)
        {
            return null;
        }

        ConstructorInfo namedPipeCtor = namedPipeConnectionInfoType.GetConstructor(new[] { typeof(int) });
        if (namedPipeCtor == null)
        {
            return null;
        }

        ParameterExpression pid = Expression.Parameter(typeof(int), nameof(pid));
        var lambda = Expression.Lambda<Func<int, RunspaceConnectionInfo>(
            Expression.New(namedPipeCtor, pid),
            new ParameterExpression[] { pid });

        return lambda.Compile();
    }
}
```

Creating a delegate drops a lot of the performance hits from using reflection (well, with some trade off in class initialization).  Also the only thing you need to call via reflection is the constructor for `NamedPipeConnectionInfo` since the base class *should* be accessible in older versions.

Note: I haven't tested this, some minor changes may be needed.

Reply:
```csharp
private static Type _namedPipeConnectionInfoType = Type.GetType("System.Management.Automation.Runspaces.NamedPipeConnectionInfo, System.Management.Automation");

private static Runspace GetRemoteRunspace(int pid)
{
    var namedPipeConnectionInfoInstance = Activator.CreateInstance(
        _namedPipeConnectionInfoType,
        pid);

    return RunspaceFactory.CreateRunspace(namedPipeConnectionInfoInstance as RunspaceConnectionInfo);
}
```

Was able to get rid of the usage of `System.Reflection`... this seems a lot smaller than what you're doing... Is it _that_ much more performant?